File: Trees/branchSums.js
class BinaryTree {
	constructor(value) {
		this.value = value;
		this.left = null;
		this.right = null;
	}

	branchSums(root, sum = 0, result = []) {
		sum += root.value;
		if (!root.left && !root.right) {
			result.push(sum);
		} else {
			if (root.left) this.branchSums(root.left, sum, result);
			if (root.right) this.branchSums(root.right, sum, result);
		}
		return result;
	}
}

// Time Complexity - O(n)
// Space Complexity - O(n)

File: Python/Easy Python/branch_sums.py
def branchSums(root):
    return branchSumsHelper(root, sum=0, list=[])


def branchSumsHelper(root, sum, list):
    sum += root.value
    if root.left == None and root.right == None:
        list.append(sum)
    else:
        if root.left:
            branchSumsHelper(root.left, sum, list)
        if root.right:
            branchSumsHelper(root.right, sum, list)
    return list

# Time Complexity: O(n)
# Space Complexity: O(n)
